osgar/drivers/rosmsg.py

```diff
@@ -252,34 +286,80 @@ def get_packet(self):
         ret, self._buf = data[:size], data[size:]        
         return ret
 
+    def slot_raw(self, timestamp, data):
+#        self._buf += data
+#        packet = self.get_packet()
+#        if packet is None:
+#            return
+#        if self.header is None:
+#            self.header = packet
+#            return
+#        self.count += 1
+#        if self.count % self.downsample != 0:
+#            return
+        packet = data  # ZMQ hack
+        if data.startswith(b'origin'):
+            s = data.split()
+            if b'ERROR' in s:
+                self.bus.publish('origin', s[1:])
+            else:
+                self.bus.publish('origin', [s[1]] + [float(x) for x in s[2:]])
+            return
+        frame_id = get_frame_id(data)
+        # TODO parse properly header "frame ID"
+        if frame_id.endswith(b'/base_link/camera_front'):  #self.topic_type == 'sensor_msgs/CompressedImage':
+            self.bus.publish('image', parse_jpeg_image(packet))
+        elif frame_id.endswith(b'/base_link/front_laser'):  #self.topic_type == 'sensor_msgs/LaserScan':
+            self.count += 1
+            if self.count % self.downsample != 0:
+                return
+            self.bus.publish('scan', parse_laser(packet))
+        elif frame_id.endswith(b'/odom'):  #self.topic_type == 'nav_msgs/Odometry':
+            __, (x, y, heading) = parse_odom(packet)
+            self.bus.publish('pose2d', [round(x*1000),
+                                        round(y*1000),
+                                        round(math.degrees(heading)*100)])
+        elif frame_id.endswith(b'/base_link/imu_sensor'):  # self.topic_type == 'std_msgs/Imu':
+            acc, rot = parse_imu(packet)
+            self.bus.publish('rot', [round(math.degrees(angle)*100) 
+                                     for angle in rot])
+            self.bus.publish('acc', [round(x * 1000) for x in acc])
+        elif frame_id.endswith(b'/clock'):
+            prev = self.timestamp_sec
+            self.timestamp_sec, self.timestamp_nsec = parse_clock(packet)
+            if prev != self.timestamp_sec:
+                self.bus.publish('sim_time_sec', self.timestamp_sec)
+
+            ms = self.timestamp_nsec//1000000
+            if self.timestamp_sec > 0 and ms % 50 == 0:  # 20Hz
```

Comment:
What is the frequency of `/clock`? Does this reliably hit 20Hz? If the frequency was strange, it wouldn't have to.

Reply:
It is 4ms, with reliable repetition (once it starts).